RecursiveExtractor/Extractors/RarExtractor.cs

```diff
@@ -89,10 +122,43 @@ public IEnumerable<FileEntry> Extract(FileEntry fileEntry, ExtractorOptions opti
             {
                 Logger.Debug(Extractor.DEBUG_STRING, ArchiveFileType.RAR, fileEntry.FullPath, string.Empty, e.GetType());
             }
-
+            var needsPassword = false;
+            try
+            {
+                needsPassword = rarArchive?.TotalUncompressSize == 0;
+            }
+            catch (Exception e)
+            {
+                needsPassword = true;
+            }
+            if (needsPassword is true)
+            {
+                var passwordFound = false;
+                foreach (var passwords in options.Passwords.Where(x => x.Key.IsMatch(fileEntry.Name)))
+                {
+                    if (passwordFound) { break; }
+                    foreach (var password in passwords.Value)
+                    {
+                        if (passwordFound) { break; }
+                        try
+                        {
+                            fileEntry.Content.Position = 0;
+                            rarArchive = RarArchive.Open(fileEntry.Content, new SharpCompress.Readers.ReaderOptions() { Password = password, LookForHeader = true });
+                            if (rarArchive.TotalUncompressSize > 0)
+                            {
+                                passwordFound = true;
```

Comment:
Why not place the break statement here to avoid another inner loop iteration?

Reply:
Done.